big refactor, much faster, 99% automatic

jmlee337 opened the pull request:
- fetch all sets using REST (no rate limits)
- set VOD URL by single batch GQL request (this makes rate limits irrelevant)
- switch to matching videos to sets, this allows us to remove the confirmation since we can know for sure a video title matches exactly one set

jmlee337 committed:
big refactor, much faster, 99% automatic

fetch all sets using REST (no rate limits)
switch to matching videos to sets, this allows us to remove the confirmation since we can know for sure a video title matches exactly one set
set VOD URL by single batch GQL request (this makes rate limits irrelevant)

jmlee337 commented:
oh, also adds `--api_key` option to set api key from the command line instead of env var

jmlee337 commented:
also also, changes sgg input to single tournament slug (checks all sets in tournament)

davisdude commented on scripts/ingest_startgg.py:
```diff
@@ -3,190 +3,196 @@
+    date = datetime.datetime.fromtimestamp(time, tz=ZoneInfo(timezone)).strftime("%Y-%m-%d")
+    phase_id_to_name = dict[str, str]()
+    for phase in tournament_response["entities"]["phase"]:
+        phase_id_to_name[phase["id"]] = phase["name"]
```
I've never seen that `dict[str, str]()` syntax for making a dict before - neat!

A more "pythonic" way to do this might be

```python
phase_id_to_name = {
    phase["id"]: phase["name"]
    for phase in tournament_response["entities"]["phase"]
}
```

davisdude commented on scripts/ingest_startgg.py:
```diff
@@ -3,190 +3,196 @@
+            gamerTags: list[str] = []
+            for key in entrant["mutations"]["participants"]:
+                gamerTags.append(entrant["mutations"]["participants"][key]["gamerTag"])
+            entrant_id_to_gamer_tags[entrant["id"]] = gamerTags
```
Same idea here - I'd prefer a dict comprehension

```python
entrant_id_to_gamer_tags = {
    entrant["id"]: [
        id["gamerTag"]
        for id in entrant["mutations"]["participants"].values()
    ]
    for entrant in group_response["entities"]["entrants"]
}
```

davisdude commented on scripts/ingest_startgg.py:
```diff
@@ -3,190 +3,196 @@
+                    "phase_name": phase_id_to_name[set["phaseId"]],
+                    "entrant_1_gamer_tags": entrant_id_to_gamer_tags[set["entrant1Id"]],
+                    "entrant_2_gamer_tags": entrant_id_to_gamer_tags[set["entrant2Id"]],
+                    "entrant_1_character_ids": set["entrant1CharacterIds"] if "entrant1CharacterIds" in set else [],
```
`Dict.get` allows you to set a fallback value

```suggestion
                    "entrant_1_character_ids": set.get("entrant1CharacterIds", [])
```

davisdude commented on scripts/ingest_startgg.py:
```diff
@@ -3,190 +3,196 @@
+            for key in entrant["mutations"]["participants"]:
+                gamerTags.append(entrant["mutations"]["participants"][key]["gamerTag"])
+            entrant_id_to_gamer_tags[entrant["id"]] = gamerTags
+        for set in group_response["entities"]["sets"]:
```
`set` is a valid python keyword - could you use something else here?

davisdude commented on scripts/ingest_startgg.py:
```diff
@@ -3,190 +3,196 @@
+                for gamer_tag in set["entrant_1_gamer_tags"] + set["entrant_2_gamer_tags"]:
+                    if normalize(gamer_tag) not in video_title_safe:
+                        all_gamer_tags_found = False
+                        break
```
```python3
all_gamer_tags_found = all([
    normalize(gamer_tag) in video_title_safe
    for gamer_tag in set["entrant_1_gamer_tags"] + set["entrant_2_gamer_tags"]
])
```

In case you can't tell, I'm a fan of list comprehensions

davisdude commented on scripts/ingest_startgg.py:
```diff
@@ -3,190 +3,196 @@
-        print(f"Processing set {set_id}")
-        try:
-            set_data = startgg_gql.get_set_data(client, set_id)
-            vod_data = get_vod_data(set_data, tournament_name, timezone)
```
Your implementation loses the ability to fetch vod ids that are already assigned. That's probably not super useful, but was the original intention of the script. Need to decide how much I want that functionality.

jmlee337 replied:
functionality restored (will fetch VOD URLs and output to json if no `--playlist-urls`)

davisdude commented on scripts/ingest_startgg.py:
```diff
@@ -3,190 +3,196 @@
+                        break
+                if all_gamer_tags_found and (normalize(set["full_round_text"]) in video_title_safe):
+                    if normalize(set["phase_name"]) in video_title_safe:
+                        phase_matching_sets.append(set)
```
This strategy makes me a bit nervous because sometimes the video may erroneously match. Some examples I've run into:

1. Doubles mixed with singles
    * e.g. `playerA vs playerB (Round)` might also match `playerA / playerC vs playerB / playerD (Doubles Round)`
2. A player's name in the tournament title
    * e.g. `playerA vs playerB (Round)` might also match `playerA vs playerC - playerB's Cool Tournament`

You can work around 2 by accepting some common string you remove from all titles for comparing.

1 is much harder if you have a single playlist for singles and doubles (which is often the case for tournaments in Iowa, at least). If you have all of doubles recorded, doing doubles first would address the issue, but I don't think that's a reliable enough method, personally.

jmlee337 replied:
we will never proceed with an ambiguous/multiple match, only when a VOD is found to match exactly one set in the tournament.

to narrow down ambiguous/multiple matches the strategy as takes an iterative approach. in the mixed doubles + singles example above, if a match is found for the doubles set it will be removed from consideration, then the singles set will be matched in the following iteration. iterating only stops once all VODs are matched or until iterating produces no further progress (same number of ambiguous/multiple matches as previous iteration)

davisdude commented on scripts/ingest_startgg.py:
```diff
@@ -3,190 +3,196 @@
+                vod_data = get_vod_data(phase_matching_sets[0], name, date, video_url)
+                if vod_data is not None:
+                    data.append(vod_data)
+                sets.remove(phase_matching_sets[0])
```
Removing data while iterating forwards through a list in Python is [unsafe](https://docs.python.org/3/tutorial/datastructures.html#looping-techniques)

> It is sometimes tempting to change a list while you are looping over it; however, it is often simpler and safer to create a new list instead.

It does seem to work fairly reliably, but AFICT python makes no such guarantees

jmlee337 replied:
this is outside the `sets` iteration

davisdude commented on scripts/ingest_startgg.py:
```diff
@@ -3,190 +3,196 @@
+        if len(current_playlist_urls) == len(next_playlist_urls):
+            print(f"videos with multiple matches: {pprint.pformat(next_playlist_urls)}")
+            break
+        current_playlist_urls = list(next_playlist_urls)
```
`next_playlist_urls` is already a list, so this feels redundant

davisdude commented on scripts/ingest_startgg.py:
```diff
@@ -3,190 +3,196 @@
+            video_title_safe = normalize(video_title)
+            phase_matching_sets = []
+            matching_sets = []
+            for set in sets:
```
I'd love for a lot of the code in this for-loop to be made into smaller functions :slightly_smiling_face:

jmlee337 replied:
I do want to make sure the algo is readable since it's a bit complicated. are there specific parts that are difficult to read?

davisdude replied:
I'd love if it looked something like this:

```python
def get_matching_sets(video_title, sets):
    # Same logic as before
    pass

def get_sets_with_vod_data(sets: list[], playlist_urls: list[tuple[str, str]]):
    set_ids_to_video_urls = {}
    unmatched_urls = []
    while len(playlist_urls) != 0:
        next_playlist_urls = []
        for video_data in playlist_urls:
            if video_data is None:
                continue
            video_title, video_url = video_data
            exact_matches, partial_matches = get_matching_sets(video_title, sets)
            if len(exact_matches) == 1:
                match = exact_matches[0]
                set_ids_to_video_urls[match["id"]] = video_url
                sets.remove(match)
            elif len(exact_matches) != 0:
                next_playlist_urls.append(video_data)
            elif len(partial_matches) == 1:
                match = partial_matches[0]
                set_ids_to_video_urls[match["id"]] = video_url
                sets.remove(match)
            elif len(partial_matches) != 0:
                next_playlist_urls.append(video_data)
            else:
                print(f"No matches found for {video_url}")
        if len(next_playlist_urls) == len(playlist_urls): # No progress made
            print("Unmatched videos: {playlist_urls}")
            # Allow interactively assigning videos?
        playlist_urls = next_playlist_urls
    return set_ids_to_video_urls

# Maybe this should be renamed, since it's also generating the JSON data, but eh
# Or the VOD assignment could/should be moved elsewhere
def set_tournament_vod_urls(slug: str, playlist_urls: list[tuple[str, str]], api_key: str):
    print(f"Setting VOD URLs for {slug}")
    name, sets, date = get_tournament_name_sets_and_timezone(slug)
    set_ids_to_video_urls = get_sets_with_vod_data(sets, playlist_urls)
    requests = [
        startgg_gql.get_set_vod_request(set_id, video_url)
        for set_id, video_url in set_ids_to_video_urls.items()
    ]
    startgg_gql.batch_set_vods(startgg_gql.get_client(api_key), requests=requests)
    return [
        vod_data
        for set_id, video_url in set_to_video.items()
        if (vod_data := get_vod_data(set_id, name, data, video_url)) is not None
    ]
```
TLDR I think at least the logic of getting matching sets can be function-ized, and the logic for creating the json struct could be moved elsewhere to reduce duplication. 

I'd also like an interactive mode to allow for goofs (for instance, the streamer uploading cut vods makes a typo in a tag), but maybe that's lower priority.

davisdude commented on scripts/ingest_startgg.py:
```diff
@@ -3,190 +3,196 @@
 
-def process_urls(startgg_urls, playlist_urls, file):
+def process_urls(slug: str, playlist_urls: list[str], file, api_key: str):
     if len(playlist_urls) == 0:
```
If playlists are required using this method, you can change `("--playlist-urls", nargs="*")` to `("playlist_urls", nargs="+")` and this check can be removed

davisdude commented on scripts/ingest_startgg.py:
```diff
@@ -3,190 +3,196 @@
+                    if normalize(gamer_tag) not in video_title_safe:
+                        all_gamer_tags_found = False
+                        break
+                if all_gamer_tags_found and (normalize(set["full_round_text"]) in video_title_safe):
```
Requiring the full round name be present is making this not work for our tournaments, where we do short round titles, e.g. GF for Grand Finals. I initially avoided including rounds at all because of how much variation there can be around video titles in general. Though I don't think it's possible to make this automated and *not* include round titles.

Moving the round text check to the next conditional made things work better for me.

jmlee337 replied:
I've moved the round check to the next conditional, but also added a alternate check for abbreviated round name in the form WR1, LR2, WQF, GF, GFR, etc

davisdude commented on scripts/ingest_startgg.py:
```diff
@@ -3,190 +3,196 @@
+                vod_data = get_vod_data(phase_matching_sets[0], name, date, video_url)
+                if vod_data is not None:
+                    data.append(vod_data)
+                sets.remove(phase_matching_sets[0])
```
```suggestion
                sets.remove(matching_sets[0])
```

jmlee337 committed:
enumerate num videos in playlist for verifiability

jmlee337 committed:
round 1 changes, mostly list comprehensions

also check for abbreviated round name

jmlee337 commented:
pushed one round of fixes/resolutions, still working on the rest

jmlee337 committed:
restore ability to just fetch VOD URLs

also one more set -> setObj

jmlee337 committed:
`--api_key` -> `--api-key` to be consistent with `--playlist-urls`

jmlee337 commented:
all comments addressed or replied

davisdude commented on scripts/ingest_startgg.py:
```diff
@@ -3,190 +3,216 @@
+    for set_id in set_id_to_video_url:
+        requests.append(startgg_gql.get_set_vod_request(set_id, set_id_to_video_url[set_id]))
+    startgg_gql.batch_set_vods(startgg_gql.get_client(api_key), requests=requests)
+    return data
```
Minor, but I'd like if this also looked through un-assigned sets at the end and added those to `data` if they have a `vodUrl` set, that way the script can produce a full JSON in 1-fewer pass. Or maybe if you're scraping a tournament you don't have mutate-permissions for, but still want the full JSON at the end.

jmlee337 replied:
done

davisdude reviewed (commented):
Looking really good. Just a few minor things. I'll also try this out tomorrow and see if I catch anything then.

davisdude commented on scripts/ingest_startgg.py:
```diff
@@ -3,190 +3,216 @@
+            matching_sets = []
+            for setObj in sets:
+                all_gamer_tags_found = all([
+                    normalize(gamer_tag) in video_title_safe
+                    for gamer_tag in setObj["entrant_1_gamer_tags"] + setObj["entrant_2_gamer_tags"]
```
I've been thinking this might should be more strict. I'm not sure of the best way to do this yet, since tags with spaces can be tough, but it might be best to split by word. A few pitfalls of this current method:

* Tags that are fully non-ASCII will always match, e.g. :sunglasses: vs A will match all sets that A is in
* Tags that are fully contained in another entrant's tag, e.g. "Player" will match all sets that "PlayerWithLongerTag" is in

This can be kicked down the road, though, since this is an existing problem. More-so this is just a reminder to me.

jmlee337 replied:
I've added some more legal ASCII characters (that are youtube filename friendly), this helps Ble$$é not match Fable, for example. We can't do anything with spaces because youtube changes some filename chars into spaces automatically so we have to ignore spaces (EG. "s-f" becomes "s f").

I would like the emphasize that if you have round/phase name included, this is much less of a problem than you're probably thinking.

jmlee337 committed:
round 2 part 1

add some more ASCII characters (which are unaffected by Youtube renaming) to `normalize`. This helps Ble$$é not match Fable, for example.

add existing VODs from leftover sets to output data when setting

collate videos from all playlists into one list, this makes the matching algorithm work better

rename some vars to be more precise/correct/stylistic

jmlee337 committed:
round 2 part 2 subdivide matching method

only set VOD URL if diff from existing
set VOD URLs in pages of 500 (sgg gql api max, it might actually be 512 now but 500 is what I have elsewhere)
--dry-run tag to skip actually setting VOD URLs
one more round of renames

jmlee337 commented:
can't reply to the comment for some reason but, subdivided the set function as requested, all comments addressed

davisdude commented:
Looks good to me! I have a few adjustments to make for handling empty brackets / unassigned things, but I can do those post-merge. Thanks a lot!
